### DISCUSSION TITLE
Add Patreon Provider

### DISCUSSION THREAD

**Discussion by dominik-sfl** [nextauthjs/next-auth#3519] [Ideas]
Just saw that there is a [Patreon API](https://docs.patreon.com/) that can be used to authenticate users (patrons). 

Is there a way to integrate this into next-auth, or is there a reason this hasn't be or shouldn't be done?

**Comment by balazsorban44** [comment]
From the link you sent, they seem to have an OAuth provider, so it should be straightforward to integrate with NextAuth.js

**Comment by Tetedeiench** [comment]
I could build a working patreon provider, despite not being an expert.

For what it's worth, here's the code I came up with. Feel free to modify it and correct it, I'm by no means an expert, but it seems to be working. 

Here's the provider configuration in api/auth/[...nextauth].js :
```    
{
      id: 'patreon',
      name: 'PATREON',
      type: 'oauth',
      version: '2.0',
      scope: 'users pledges-to-me my-campaign',
      params: { grant_type: 'authorization_code' },
      authorizationUrl: `${process.env.PATREON_AUTHORIZE_URL}?response_type=code`,
      accessTokenUrl: `${process.env.PATREON_TOKEN_URL}`,
      profileUrl: `${process.env.PATREON_PROFILE_URL}`,
      profile: (profile) => {
        return {
          id: profile.data.id,
          name: profile.data.attributes.full_name,
          email: profile.data.attributes.email
        }
      },
      clientId: process.env.PATREON_CLIENT_ID,
      clientSecret: process.env.PATREON_CLIENT_SECRET
}
```

I like having my parameters in an ENV file, and here they are : 
```
PATREON_AUTHORIZE_URL=https://www.patreon.com/oauth2/authorize
PATREON_TOKEN_URL=https://www.patreon.com/api/oauth2/token
PATREON_PROFILE_URL=https://www.patreon.com/api/oauth2/api/current_user
PATREON_CLIENT_ID=<Your client id>
PATREON_CLIENT_SECRET=<Your client secret>
```

**Comment by Tetedeiench** [comment]
The above configuration is for NextAuth v3 and patreon OAuth2 v1 . Kind of complex with all those version numbers.

Here's a working one for NextAuth v4 and Patreon OAuth2 v2, fully up to date at the time I'm writing this : 
```    
{
      id: 'patreon',
      name: 'PATREON',
      type: 'oauth',
      version: '2.0',
      authorization: { 
        url : `${process.env.PATREON_AUTHORIZE_URL}?response_type=code`,
        params: { scope: 'identity identity[email] identity.memberships', grant_type: 'authorization_code' }
      },
      token : {
        url : `${process.env.PATREON_TOKEN_URL}`,
      },
      userinfo : {
        url: `${process.env.PATREON_PROFILE_URL}`
      },      
      profile: (profile) => {
        return {
          id: profile.data.id,
          name: profile.data.attributes.full_name,
          email: profile.data.attributes.email,
          provider: 'PATREON'
        }
      },
      clientId: process.env.PATREON_CLIENT_ID,
      clientSecret: process.env.PATREON_CLIENT_SECRET
    }
```

The env variables are the same as above. Enjoy !

**Reply by balazsorban44** [reply]
Care to open a PR adding this to the built-in ones? :slightly_smiling_face:

**Reply by Tetedeiench** [reply]
> Care to open a PR adding this to the built-in ones? 🙂

Hi,

I'm not an expert in JS dev, but NextAuth has helped me so much I'm giving it a shot - make sure you thoroughly review what I'm doing !

Forked your repo, created a branch from main as instructed, implemented the provider, tested it, it works fine. 

I wanted to write the doc, but the "www" folder is missing from main (it is present in canary). Can you help me there please ?

Once I know this, i'll finish the PR and submit it for review :)

**Reply by balazsorban44** [reply]
Awesome! Just open the PR against the `main` branch, please.

Docs should be placed at https://github.com/nextauthjs/docs/tree/main/docs/providers

**Reply by Tetedeiench** [reply]
Did the two PR (for the provider and doc). Feel free to review what i've done :) Thanks !
